domain/mempool/mempool_utxoset.go

```diff
@@ -0,0 +1,115 @@
+package mempool
+
+import (
+	"github.com/kaspanet/kaspad/app/appmessage"
+	"github.com/kaspanet/kaspad/domain/blockdag"
+	"github.com/kaspanet/kaspad/util"
+	"github.com/kaspanet/kaspad/util/daghash"
+	"github.com/pkg/errors"
+)
+
+func newMempoolUTXOSet(dag *blockdag.BlockDAG) *mempoolUTXOSet {
+	return &mempoolUTXOSet{
+		transactionByPreviousOutpoint: make(map[appmessage.Outpoint]*util.Tx),
+		poolUnspentOutputs:            make(map[appmessage.Outpoint]*blockdag.UTXOEntry),
+		dag:                           dag,
+	}
+}
+
+type mempoolUTXOSet struct {
+	transactionByPreviousOutpoint map[appmessage.Outpoint]*util.Tx
+	poolUnspentOutputs            map[appmessage.Outpoint]*blockdag.UTXOEntry
+	dag                           *blockdag.BlockDAG
+}
+
+func (mpus *mempoolUTXOSet) utxoEntryByOutpoint(outpoint appmessage.Outpoint) (entry *blockdag.UTXOEntry, isInPool bool, exists bool) {
+	entry, exists = mpus.dag.GetUTXOEntry(outpoint)
+	if !exists {
+		entry, exists := mpus.poolUnspentOutputs[outpoint]
+		if !exists {
+			return nil, false, false
+		}
+		return entry, true, true
+	}
+	return entry, false, true
+}
+
+// addTx adds a transaction to the mempool UTXO set. It assumes that it doesn't double spend another transaction
+// in the mempool, and that its outputs doesn't exist in the mempool UTXO set, and returns error otherwise.
+func (mpus *mempoolUTXOSet) addTx(tx *util.Tx) error {
+	msgTx := tx.MsgTx()
+	for _, txIn := range msgTx.TxIn {
+		if existingTx, exists := mpus.transactionByPreviousOutpoint[txIn.PreviousOutpoint]; exists {
+			return errors.Errorf("outpoint %s is already used by %s", txIn.PreviousOutpoint, existingTx.ID())
```

Comment:
I'd change `used` -> `spent`.
Or even be harsher "tx {txID} is double-spending outpoint {blabla} that is also spent transaction {otherTxID} already in mempool".